src/Microsoft.Health.Fhir.TemplateManagement/Overlay/OverlayFileSystem.cs

```diff
@@ -6,130 +6,155 @@
-
-        public OCIFileLayer ReadMergedOCIFileLayer()
+        public async Task<OciFileLayer> ReadOciFileLayerAsync()
         {
-            var filePaths = Directory.EnumerateFiles(WorkingFolder, "*.*", SearchOption.AllDirectories)
-                .Where(f => !string.Equals(GetTopDirectoryPath(Path.GetRelativePath(WorkingFolder, f)), Constants.HiddenImageFolder));
+            var filePaths = Directory.EnumerateFiles(_workingFolder, "*.*", SearchOption.AllDirectories)
+                .Where(f => !string.Equals(GetTopDirectoryPath(Path.GetRelativePath(_workingFolder, f)), Constants.HiddenImageFolder));
 
             Dictionary<string, byte[]> fileContent = new Dictionary<string, byte[]>() { };
             foreach (var oneFile in filePaths)
             {
-                fileContent.Add(Path.GetRelativePath(WorkingFolder, oneFile), File.ReadAllBytes(oneFile));
+                fileContent.Add(Path.GetRelativePath(_workingFolder, oneFile), await File.ReadAllBytesAsync(oneFile));
             }
 
-            OCIFileLayer fileLayer = new OCIFileLayer() { FileContent = fileContent };
+            OciFileLayer fileLayer = new OciFileLayer() { FileContent = fileContent };
             return fileLayer;
         }
 
-        public void WriteMergedOCIFileLayer(OCIFileLayer oneLayer)
+        public async Task WriteOciFileLayerAsync(OciFileLayer oneLayer)
         {
             EnsureArg.IsNotNull(oneLayer, nameof(oneLayer));
 
-            Directory.CreateDirectory(WorkingFolder);
+            Directory.CreateDirectory(_workingFolder);
             foreach (var oneFile in oneLayer.FileContent)
             {
-                var filePath = Path.Combine(WorkingFolder, oneFile.Key);
+                var filePath = Path.Combine(_workingFolder, oneFile.Key);
                 var directory = Path.GetDirectoryName(filePath);
                 Directory.CreateDirectory(directory);
-                File.WriteAllBytes(Path.Combine(WorkingFolder, oneFile.Key), oneFile.Value);
+                await File.WriteAllBytesAsync(Path.Combine(_workingFolder, oneFile.Key), oneFile.Value);
             }
         }
 
-        public List<OCIArtifactLayer> ReadImageLayers()
+        public async Task<List<ArtifactBlob>> ReadImageLayersAsync()
         {
-            return ReadOCIArtifactLayers(WorkingImageLayerFolder);
+            return await ReadOciArtifactLayersAsync(_workingImageLayerFolder);
         }
 
-        public void WriteImageLayers(List<OCIArtifactLayer> imageLayers)
+        public async Task WriteImageLayersAsync(List<ArtifactBlob> imageLayers)
         {
             EnsureArg.IsNotNull(imageLayers, nameof(imageLayers));
 
-            ClearFolder(WorkingImageLayerFolder);
+            Directory.CreateDirectory(_workingImageLayerFolder);
+
+            // Used to label layers sequence. In this version, only two layers will be generated.
+            // Should not exceed 9 in the future.
+            var layerNumber = 1;
```

Comment:
It might be better to keep this into a constant as well, so users can notice that our layers start from index 1 more quickly.